### DISCUSSION TITLE
Sending back events from a promise actor

### DISCUSSION THREAD

**Discussion by Ghirigoro** [statelyai/xstate#4838] [Q&A]
Sending back from a callback actor is really straightforward - a nicely typed `sendBack` function is provided for you by XState and you're good to go. It's not clear to me exactly what the process is for sending back with a promise actor. As far as I can tell the way you would do this is by getting the parent and then calling send on it. For example:
```ts
const parent = system.get(parentId);
if (parent === undefined) throw new Error("Parent machine not found");
// parent is `any` so we have to cast
parent.send({ type: "response", response: null } as MyEventType);
```
Managing ids and a loss of typesafety doesn't seem right for such a basic task, especially since similar tasks have a clean API for it. What am I missing? Looking at the docs the only thing I see is a mention about passing in a reference to the parent using the input, but its not clear to me how one would do that. 

Any help would be appreciated. Thanks

**Comment by davidkpiano** [comment]
> Looking at the docs the only thing I see is a mention about passing in a reference to the parent using the input, but its not clear to me how one would do that.

This is how you would do that:

```ts
const someMachine = createMachine({
  // ...
  invoke: {
    src: promiseLogic,
    input: ({ self }) => ({ parent: self })
  }
});


const promiseLogic = fromPromise(({ input }: {
  input: { parent: ActorRefFrom<typeof someLogic> }
}) => {
  // ...
  input.parent.send({/* ... */}); // Should be strongly-typed
  // ...
})
```

**Reply by codingedgar** [reply]
Could `fromPromise` have `sendBack` just like `fromCallback`?

```ts
      const resolvedPromise = Promise.resolve(
        promiseCreator({
          input: state.input!,
          system,
          self,
          signal: controller.signal,
          emit,
          sendBack: (event) => {
            if (self.getSnapshot().status === 'stopped') {
              return;
            }
            if (self._parent) {
              system._relay(self, self._parent, event);
            }
          },
        })
      );
```

**Comment by ayush-sd** [comment]
When I try to do something similar I get these circular reference errors. I understand that the problem here has to do with referencing `chatMachine` inside its own definition but I am unable to figure out how to work around this, could you please help? @davidkpiano 

https://stackblitz.com/edit/vitejs-vite-rzerqvhg?file=src%2Fmain.ts

### Screenshot
<img width="1330" alt="image" src="https://github.com/user-attachments/assets/86bfd163-82b6-47bf-9f3f-bbe1c8a92a5c" />

**Reply by with-heart** [reply]
This is unfortunately just a side effect of how TypeScript works. Since the type of `chatMachine` is derived from its config (including its `actors`), any references to the type of `chatMachine` from within itself creates a circular relationship.

I think the easiest way to avoid circular references here would be to create a separate actor type which contains only the type information we need from it. Since we're only using `input.parent.send` to send events back to the parent machine (I think?), the only type information we need is the parent's events.

We need to do 2 things:

1. define the events type outside of `chatMachine` (so we can reference them in `chatMachine` _and_ our new actor type)
2. create the new actor type

```ts
import { type ActorRef, type AnyMachineSnapshot } from 'xstate'

type ChatEvent =
  | { type: 'message.send' }
  | { type: 'promptInput.update'; prompt: string }
  | { type: 'markdown.append'; markdown: string }

type ChatInputActor = ActorRef<AnyMachineSnapshot, ChatEvent>
// we use `AnyMachineSnapshot` because we don't care about the machine's context, etc. when using `input.parent`
```

Now that we have this independent `ChatInputActor` type, we can safely use it within `chatMachine` without causing circular references.

Here's your example updated to use this new setup: https://stackblitz.com/edit/vitejs-vite-jwjejydr?file=src%2Fmain.ts
